UCSC Extension: Bioinformatics Tools, Databases, and Methods
============================================================

My notes for UCSC Extension's Bioinformatics Tools course taught by Lee Kozar.
This course is an introduction to the public domain tools and databases like
GenBank, PBD, and BLAST.

The notes are in a question/answer format.  They go better with the slides from
lectures.

Bioinformatics Databases
========================

## What database do you search for info about introns in a gene?
## What database do you search for genetic diseases caused by a gene?
## What database do you search for gene therapy information with a gene?
## What database do you search for info about polymorphic variation in a gene?
## Can you use SRS to search sequence databases?
## Can you use MeSH to search sequence databases?
## Why are there so many bioinformatics databases?
## If you have a sequence in GenBank format, why would you convert it to FASTA?
## Can a FASTA file hold more than one sequence?
## How does a computer know when a sequence starts/ends in a GenBank file?
## Can a GenBank file hold more than one sequence?
## Can a sequence only file hold more than one sequence?  Why/why not?
## What is the EST database used for?
## Why would you place "limits" on your search of the GenBank database?
## What info is lost on conversion of GenBank sequence to FASTA?
## Can that information be put back if you convert back to GenBank?
## What information is lost when a SwissProt file is converted to FASTA format?

Literature Searching
====================

## Why is MeSH useful?
## What are the advantages and disadvantages of a controlled vocabulary?
## What database do you search to find a gene given the discoverer's last name?
## What searches can't be done at the PubMed/Medline database site?
## What is contained in the annotation part of the sequence?
## What is more important, annotation or sequence?  Why?
## What does "parsing" do to a database?
## Why are relational databases better than flat file databases?
## What database/terms do you use to find human growth hormone protein sequence?
## How does SRS differ from NCBI Search tools?
## Why would you place "limits" on your search of the PubMed database?

Statistics in Bioinformatics
============================

## Why is probability and statistics important for bioinformatics?
## How often would you find any dinucleotide sequence in a random sequence?
## How often would you find any tri-nucleotide sequence in a random sequence?
## Given a nucldeotide sequence with random distribution, what can you conclude?
## Given a non-random distribution, what does it tell you?
## When would you use a Hidden Markov Model to analyze a sequence?
## What is information theory used for in bioinformatics?
## How often do you expect a single nucleotide to appear in a random sequence?
## Is a sequence random if the four nucleotides appear in same frequency?
## What further tests would you do to make sure the sequence is random?

To answer this, come up with some sequences that had the same distribution
of nucleotides but was not random.

## How often would you find the cut site for HindIII in a random sequence?

To answer this, you need to find information about the HindIII cut site and its
length.

## What is a sliding window technique and why is it useful?
## Why is leucine the most common amino acid?
## What is a codon preference table?  When would you use one?
## Can the selection of a codon be used to regulate transcription rate?
## Can the selection of a codon be used to regulate gene expression?
## If you shuffle a sequence, will the single nucleotide frequencies change?
## You shuffle a sequence and di-nucleotide frequencies stay constant.  Why?
## How is a bioinformatics experiment designed?
## Where does the data from a bioinformatics experiment come from?
## How do you know if a bioinformatics experiment worked?
## What is Bayes Equation?  Why is it so useful?
